File: morning_exercises/day5.py
def isValid(input: str)->bool:
    openings = []
    mappings = {
        '}':'{',
        ']':'[',
        ')':'('
    }
    
    for c in input:
        if c in mappings.values(): openings.append(c)

        else:
            if len(openings)==0: return False

            if (openings[-1]!=mappings[c]): return False

            openings.pop()

    return len(openings)==0

print(isValid("{[()]}"))

File: morning_exercises/day5.ts
function isValid(input: string): boolean {
    const openings: string[] = [];
    const mappings = new Map<string, string>([
        [']', '['],
        ['}', '{'],
        [')', '(']
    ]);

    for (const c of input) {
        if(c==='{' || c==='[' || c==='(') openings.push(c);
        else {
            if (openings.length===0) return false;

            if (openings.at(-1)!=mappings.get(c)) return false;

            openings.pop();
        }
    }
    return (openings.length===0);
}

console.log(isValid("{{[]}}"))